transformer_engine/common/comm_gemm_overlap/userbuffers/userbuffers.cu

```diff
@@ -2597,30 +2598,54 @@ void reset_counters(void *atomic_ptr, int num_chunks, bool allgather, cudaStream
   reset_counters_kernel<<<grid, block, 0, stream>>>(atomic_ptr, num_chunks, allgather);
 }
 
-template <typename fp8type>
+template <typename fp8type, int nvec>
 __global__ void __launch_bounds__(MAX_THREADS / 4)
     reduce_fp8_in_bf16_out_cuda(void *inputs, void *output, const float *scale,
-                                const int num_inputs, const int input_size) {
-  const size_t tid = threadIdx.x + blockDim.x * blockIdx.x;
+                                const int num_inputs, const int input_size,
+                                const int num_aligned_elements_per_input,
+                                const int tot_input_size) {
   fp8type *inputs_fp8 = reinterpret_cast<fp8type *>(inputs);
-  float accum_buf = static_cast<float>(inputs_fp8[tid]) * (*scale);
+  half_dtype *output_half = reinterpret_cast<half_dtype *>(output);
+
+  transformer_engine::VectorizedLoader<fp8type, nvec, true> loader(inputs_fp8, tot_input_size);
+  transformer_engine::VectorizedStorer<half_dtype, nvec, true> storer(output_half, input_size);
+
+  const size_t tid = threadIdx.x + blockDim.x * blockIdx.x;
```

Comment:
Do we handle the case where the block size doesn't neatly divide the input size? Maybe we can fix with something like:
```suggestion
  const size_t tid = threadIdx.x + blockDim.x * blockIdx.x;
  if (tid > num_aligned_elements_per_input) {
    return;
  }
```
Alternatively, we can change how we configure the CUDA blocks:
```c++
  size_t num_threads = MAX_THREADS / 4;
  assert(num_aligned_elements_per_input % num_threads == 0);
  size_t num_blocks = num_aligned_elements_per_input / num_threads;
  dim3 block(num_threads);
  dim3 grid(num_blocks);
```
Same issue in `reduce_bf16_cuda`.